app/api/contests/set_additional_time.go

```diff
@@ -57,59 +61,132 @@ import (
+		}
+		service.MustNotBeError(err)
+
+		setAdditionalTimeForGroupInContest(store, groupID, itemID, contestInfo.ContestParticipantsGroupID,
+			contestInfo.DurationInSeconds, seconds)
+		return nil
+	})
+	if apiError != service.NoError {
+		return apiError
+	}
+	service.MustNotBeError(err)
 
 	render.Respond(w, r, service.UpdateSuccess(nil))
 	return service.NoError
 }
 
-func (srv *Service) setAdditionalTimeForGroupInContest(groupID, itemID, seconds int64) {
-	service.MustNotBeError(srv.Store.InTransaction(func(store *database.DataStore) error {
-		groupContestItemStore := store.GroupContestItems()
-		scope := groupContestItemStore.Where("group_id = ?", groupID).Where("item_id = ?", itemID)
-		found, err := scope.WithWriteLock().HasRows()
-		service.MustNotBeError(err)
-		if found {
-			service.MustNotBeError(scope.UpdateColumn("additional_time", gorm.Expr("SEC_TO_TIME(?)", seconds)).Error())
-		} else if seconds != 0 {
-			service.MustNotBeError(groupContestItemStore.Exec(
-				"INSERT INTO groups_contest_items (group_id, item_id, additional_time) VALUES(?, ?, SEC_TO_TIME(?))",
-				groupID, itemID, seconds).Error())
-		}
-		return nil
-	}))
+func (srv *Service) getParametersForSetAdditionalTime(r *http.Request) (itemID, groupID, seconds int64, apiError service.APIError) {
+	itemID, err := service.ResolveURLQueryPathInt64Field(r, "item_id")
+	if err != nil {
+		return 0, 0, 0, service.ErrInvalidRequest(err)
+	}
+	groupID, err = service.ResolveURLQueryPathInt64Field(r, "group_id")
+	if err != nil {
+		return 0, 0, 0, service.ErrInvalidRequest(err)
+	}
+	seconds, err = service.ResolveURLQueryGetInt64Field(r, "seconds")
+	if err != nil {
+		return 0, 0, 0, service.ErrInvalidRequest(err)
+	}
+	const maxSeconds = 838*3600 + 59*60 + 59
+	// 838:59:59 is the maximum possible TIME value in MySQL
+	if seconds < -maxSeconds || maxSeconds < seconds {
+		return 0, 0, 0, service.ErrInvalidRequest(fmt.Errorf("'seconds' should be between %d and %d", -maxSeconds, maxSeconds))
+	}
+	return itemID, groupID, seconds, service.NoError
+}
+
+func setAdditionalTimeForGroupInContest(
+	store *database.DataStore, groupID, itemID, participantsGroupID, durationInSeconds, additionalTimeInSeconds int64) {
+	groupContestItemStore := store.GroupContestItems()
+	scope := groupContestItemStore.Where("group_id = ?", groupID).Where("item_id = ?", itemID)
+	found, err := scope.WithWriteLock().HasRows()
+	service.MustNotBeError(err)
+	if found {
+		service.MustNotBeError(scope.UpdateColumn("additional_time",
+			gorm.Expr("SEC_TO_TIME(?)", additionalTimeInSeconds)).Error())
+	} else if additionalTimeInSeconds != 0 {
+		service.MustNotBeError(groupContestItemStore.Exec(
+			"INSERT INTO groups_contest_items (group_id, item_id, additional_time) VALUES(?, ?, SEC_TO_TIME(?))",
+			groupID, itemID, additionalTimeInSeconds).Error())
+	}
+
+	service.MustNotBeError(store.Exec("DROP TEMPORARY TABLE IF EXISTS total_additional_times").Error())
+	service.MustNotBeError(store.Exec(`
```

Comment:
This request probably needs some explanations (comments), especially as there are many kind of groups and it is very hard to follow.
For instance, indicating after L157 that, at this point, we have all groups participating to the contest and which are descendant of the group on which we change additional time (so, that will be affected)

Reply:
fixed